tutorials/snapcraft/snap-a-nodejs-app.md

```diff
@@ -0,0 +1,666 @@
+
+### Adding some apps command
+
+Before exposing your server as a service, we strongly recommend you to declare it as a simple snap application. This enables you to easily debug your service and iterate, especially when adding confinement. So let’s roll in!
+
+If you remember from previous tutorials, exposing a command is easy:
+
+
+```bash
+apps:
+  node-service:
+    command: bin/chuck-norris-app
+```
+
+Note that we reference here the starting binary file generated by `package.json`.
+
+Let’s rebuild it and use `snap try` to install the local snap before running the command!
+
+
+```bash
+$ snapcraft prime
+Skipping pull webserver (already ran)
+Skipping build webserver (already ran)
+Skipping stage webserver (already ran)
+Skipping prime webserver (already ran)
+$ snap try prime/ --devmode
+chuck-norris-webserver 1.0.0 mounted from[…]/prime
+$ chuck-norris-webserver.node-service
+events.js:141
+      throw er; // Unhandled 'error' event
+      ^
+
+Error: listen EACCES 0.0.0.0:80
+    at Object.exports._errnoException (util.js:870:11)
+    at exports._exceptionWithHostPort (util.js:893:20)
+    at Server._listen2 (net.js:1221:19)
+    at listen (net.js:1270:10)
+    at Server.listen (net.js:1366:5)
+    at Object.<anonymous> (/snap/chuck-norris-webserver/x1/lib/node_modules/chuck-norris-app/main.js:71:8)
+    at Module._compile (module.js:409:26)
+    at Object.Module._extensions..js (module.js:416:10)
+    at Module.load (module.js:343:32)
+    at Function.Module._load (module.js:300:12)
+```
+
+Ah, a failure! We are getting a javascript stacktrace. Looking a little bit more closely, we see that the exact error is: `Error: listen EACCES 0.0.0.0:80`. Indeed, unprivileged users on a Linux system can’t listen to **ports < 1024**. We then need to run the command as the root user:
+
+
+
+```bash
+$ sudo chuck-norris-webserver.node-service 
+Server listening on: http://localhost:80
+```
+
+You can now point your browser to [http://localhost] and enjoy some Chuck Norris quotes! All debug outputs will be visible on the command line prompt. Press Ctrl+C to exit it.
+Having a command while you are working on your snap helps you greatly in inspecting what’s going on in your snap. This is why we strongly recommend that approach.
+
+Also, that means to listen to that port 80, we’ll need the service to run as a privileged user. The good news is that in the snap world, thanks to confinement, all services are running as root while being safe!
+It’s something to keep in mind while you iterate on a command which is going to be turned into a service later on: run them as root to reproduce the exact running context!
+
+It’s now high time to get rid of this devmode thing and add some confinement to your snap, let’s do this right away!
+
+ 
+## Turning on confinement!
+Duration: 5:00
+
+
+```bash
+Lost or starting from here?
+Check or download [here-3] to see what your current directory should look like.
```

Comment:
Link is broken, should be "[here](https://github.com/ubuntu/snap-tutorials-code/tree/master/build-a-nodejs-service/step3)"